### ISSUE TITLE
tn-cli, GRPC - problems to create group topics, join them and unsubscribe

### ISSUE THREAD

**Issue by sanmiguel2019** [tinode/chat#574]
#294 # Subject of the issue

These problems I see in the webapp when sending requests from GRPC. Have not tested it on mobile apps but I guess there will be the same problem. I guess Tinode is incorrectly communicating with clients when request is sent `on_behalf_of`

1) When joining a group topic (using GRPC, on_behalf_of), user gets `modegiven` permission `N` despite to the fact that group has the following permissions: `{"Auth":"JRWPD","Anon":"N"}` Acually, user joins a group and at the same time gets blocked. It's necessary to send another request `set` on behalf of the group topic owner to set correct permissions. I even tried to change "Anon":"N" to "Anon":"JRWPD" but it does not help. There are no problems when joining the same group topic from the webapp. 

2) New created group topics dont appear in the list of topics (contacts) in the webapp without page reload. At the same time new created p2p topics immediately appear in the list of topics.

3) It's not possible to unsubscribe from a group topic by deleting the topic or by sending `leave` request with `unsub:true`.

4) It's not possible to kick out user from a topic if user recently has opened the same topic (attached to it).

#### Server-side
SDK: tinodejs/0.16.8
label_server 0.16 mysql:v0.16.8

#### Client-side
Tinode command line client. Version 1.5.3/0.16.7; 
gRPC/1.33.1; 
Python 2.7.16.

### Problem N 1

Root user logs in and subscribes user `usrttPQMt5Q9aw` to the group topic `grpFNKkZ4uzy1s`
```

2020/11/29 13:45:52 grpc in: login:<id:"597391" scheme:"token" secret:"\266\323\3202\336P\365\254\365\251\313_\036\000\001\000\001\000\033%\035\214\320\220\333#\330\245\025\005\242\340\205\322\357\210FiQ\347\312\352\004\020\311\000F\336\322T" > auth_level:ROOT  XfJDDVktbcE
2020/11/29 13:45:52 grpc in: sub:<id:"284955" topic:"grpFNKkZ4uzy1s" > on_behalf_of:"usrttPQMt5Q9aw"  XfJDDVktbcE
```

This request is successful but user `usrttPQMt5Q9aw` gets modegiven permission `N`.

To set correct permissions it's necessary to send another, `set` request on behalf of the topic owner `usrmohy3Tcxp1A`

`2020/11/29 14:01:45 grpc in: set:<id:"167330" topic:"grpFNKkZ4uzy1s" query:<sub:<user_id:"usrttPQMt5Q9aw" mode:"JRWPD" > > > on_behalf_of:"usrmohy3Tcxp1A"  c4pYkm4kWp4`

Now user `usrttPQMt5Q9aw` has correct permissions but at the same time the topic does not appear in the contacts list of user `usrttPQMt5Q9aw`

### Problem N 2

If I create a new group topic (as a root user) on behalf of another user and that user currently is using the webapp, the new topic does not apperar in the list of contacts. It only will appears on page reload or on reconnect if the webapp loses connection to the server. At the same time new p2p topics created on behalf of another user, immediately appear in the list of conversations.
**Repoducing:** login into webapp as user A. Send a GRPC request as root user to create a group topic on behalf of user A. Check the webapp. You will not see new group topic in the contacts list.

**GRCP - server log:**

grpc in: sub:<id:"561038" topic:"new204137" set_query:<desc:<default_acs:<auth:"JRWP" anon:"N" > public:"{\"fn\":\"group name\",\"photo\":{\"type\":\"png\",\"data\":\"iVBORw0KGgoAAAANSUhEUgAAAAMAAAADCAIAAADZSiLoAAAACXBIWXMAAAsSAAALEgHS3X78AAAAEElEQVR42mP4n8YAQQxYWADFWAyODQsbrwAAAABJRU5ErkJggg==\"}}" > sub:<mode:"JRWPO" > > get_query:<what:"data sub desc" sub:<user:"usrmohy3Tcxp1A" topic:"new204137" limit:20 > > > on_behalf_of:"usrmohy3Tcxp1A"  NAMJtYRMACg

**GRCP - webapp client log:**

[14:49:20.157] in: {"pres":{"topic":"me","src":"grp32gJeumxMVU","what":"on"}} 
[14:49:24.160] in: {"pres":{"topic":"me","src":"grp32gJeumxMVU","what":"off"}} 

### Problem N 3

Topic owner can delete a group topic with all the subscriptions but user who joins a group topic can not delete subscription, can not leave the topic.  Only it´s possible to hide the topic. **Reproducing:** try to delete some group topic in the webapp. I tried it also fromm GRPC. The topic will disappear from the list of conversations but subscription will stay in the database, only DB field  value `deletedat `will be updated. If user decides to delete the topic, the subscription must be deleted. If not, the label `Delete` should be renamed to `Hide` but I don´t understand why to keep garbage in the database. If user unsubscribes, subscription entry should be deleted.

`leave` message with `unsub: true` to the server does not work either. It does not remove the subscription, only updates `deletedat` value in the database:

`2020/11/29 11:51:15 grpc in: leave:<id:"309053" topic:"grpPqlIeuHtu3g" unsub:true > on_behalf_of:"usrttPQMt5Q9aw"  5P7KxBkzKE0`

as you can see, the request contains `unsub:true` but it does not unsubscribe user, subscription entry stays in the database.

### Problem N 4

I am trying (as a root user) to kick out user from a group topic without blocking it ( I use GRPC). It's not possible for 2 reasons:

a) if I send `leave` message to the server on behalf of user that I want to kick out, this group topic will not disappear from the conversations list.

`leave` request sent from GRPC does not even hide the topic:

**Server log:**

```
2020/11/29 15:48:42 grpc in: login:<id:"500093" scheme:"token" secret:"\266\323\3202\336P\365\254\365\251\313_\036\000\001\000\001\000\033%\035\214\320\220\333#\330\245\025\005\242\340\205\322\357\210FiQ\347\312\352\004\020\311\000F\336\322T" > auth_level:ROOT  zZznltgUVfQ
2020/11/29 15:48:42 grpc in: sub:<id:"107132" topic:"grpFNKkZ4uzy1s" > on_behalf_of:"usrttPQMt5Q9aw"  zZznltgUVfQ
2020/11/29 15:48:42 grpc in: leave:<id:"244924" topic:"grpFNKkZ4uzy1s" unsub:true > on_behalf_of:"usrttPQMt5Q9aw"  zZznltgUVfQ
```

**Webapp client log:**

```
[15:48:21.279] in: {"pres":{"topic":"me","src":"grpFNKkZ4uzy1s","what":"acs","dacs":{"want":"JRWPD"}}} 
[15:48:21.282] in: {"pres":{"topic":"me","src":"grpFNKkZ4uzy1s","what":"on"}} 
```

Topic does not disappear from the contacts list.

b) if user that I want to kick out currently is typing ( and is attached) in this topic, he will even not notice anything - the topic will not disappear and when he sends a new message, he will be automatically subscribed again.

### Expected behaviour
- when user joins a topic he should get correct permissions instead of `N`
- new created group topics should immediately appear in he contacts list
- `delete hard` of a group topic message to the server should delete the subscription instead of updating `deletedat `value in the database
- `leave` with `unsub: true` message to the server should delete the subscription instead of updating `deletedat `value in the database

**Comment by ssallari** [comment]
I have the same issue. I want to create group and join users to this group by tn-cli. tn-cli macro code run from a php program, but i can not join users to group. my problem is to find user. we have another db that store our users.
.must sub fnd 
.must set fnd --tags=i1009usr 
.must $meta get fnd --sub 

can you give me some more example about find user in tinode to solve this problem or any code run on the server to find user and join to group.
thanks a lot

**Comment by or-else** [comment]
> I have the same issue

No you do not have the same issue. 

Please read the conversations at https://groups.google.com/d/forum/tinode
If you still have not found the answer after reading the conversation there, please ask your question.

**Comment by sanmiguel2019** [comment]
Updates of:

### Problem N 2

In the scenario described abowe one of 2 users was root. Now I send a GRPC request as user A (root) to create a p2p topic between users B and C, on behalf of user B.

```
auth_level:ROOT  iwNF5AvS60A
2020/12/01 21:57:23 grpc in: sub:<id:"156112" topic:"usrl7tzvVEjGvI" > on_behalf_of:"usrmohy3Tcxp1A"  iwNF5AvS60A
2020/12/01 21:57:23 topic[p2pmohy3Tcxp1CXu3O9USMa8g] subscription failed topic access denied; user is banned, sid=iwNF5AvS60A
```
Appears error message "user is banned" but Tinode creates topic in the database and subscribtions.
Below are permissions in the database when I do this from webapp and from GRPC:

![tinodep2p](https://user-images.githubusercontent.com/54183931/100802952-31903980-3422-11eb-9b8a-0f606bad91d5.png)

Permissions of the new created p2p topic:` {"Auth":"N","Anon":"N"}`

Webapp´s config.js file contains:

```
export const DEFAULT_P2P_ACCESS_MODE = 'JRWPS';
export const NEW_GRP_ACCESS_MODE = 'JRWPSAO';
```

but how it´s possible to send access mode using GRPC? I think these settings should be in tinode.conf file, not in the config files of the clients. I tried:

```
auth_level:ROOT  lHVOvpVv8wk
2020/12/01 21:59:28 grpc in: sub:<id:"573572" topic:"usrl7tzvVEjGvI" set_query:<desc:<default_acs:<auth:"JRWPA" anon:"N" > public:"{\"fn\":\"\",\"photo\":{\"type\":\"\",\"data\":\"\"}}" > sub:<mode:"JRWPS" > > get_query:<what:"data sub desc" sub:<user:"usrl7tzvVEjGvI" topic:"usrl7tzvVEjGvI" limit:20 > > > on_behalf_of:"usrmohy3Tcxp1A"  lHVOvpVv8wk
2020/12/01 21:59:28 topic[p2pmohy3Tcxp1CXu3O9USMa8g] subscription failed topic access denied; user is banned, sid=lHVOvpVv8wk
```

but it does not affect anything. Both users get `modewant` and `modegiven` values as in the image above.


**p2p topics in the webapp**

I am testing Tinode for a long time and don´t know if this problem in the webapp existed earlier or it´s new and appeared because of some updates. New p2p topics don´t appear in the **Webapp**.

**New p2p topics don´t appear to the second person without page reload. They don´t appear even if messages are sent.**

Reproducing:

- Log in into sandbox.tinode.co as user B
- Log in into sandbox.tinode.co as user C from a different device.
- As user B search for user C using the search function by ID. 
- Send a message to user C.

**Result:** new p2p topic will not appear in the contacts list of user C even after user B sends messages to user C. Created p2p topic only will appear when user C reloads the page.

This is webapp´s log of user C after user B creates a p2p topic with him and sends a message. Only 1 line:

`[22:25:40.388] in: {"pres":{"topic":"me","src":"usrmohy3Tcxp1A","what":"msg","seq":1,"act":"usrmohy3Tcxp1A"}} `

**Group topics in the webapp**

If user B creates a group and adds user C to the group. This group will appear in the contacts of user C. At the same time when user B sends a message to this group, user C will not see new messages badge for this group. It will appear only after page reload.

And as i wrote in my first post, when I create a group using GRPC, new group even don´t appear in added user contacts without reloading the page.

I have not tested and don´t know if there problems exists in mobile clients.

**Comment by or-else** [comment]
Problem 2 is fixed

**Comment by or-else** [comment]
**Problem 1:**

> This request is successful but user usrttPQMt5Q9aw gets modegiven permission N.

I'm sure that's because you set `DefAcs.Auth` to `N`.

>  Now user usrttPQMt5Q9aw has correct permissions but at the same time the topic does not appear in the contacts list of user usrttPQMt5Q9aw

I cannot reproduce. It's probably fixed by the same https://github.com/tinode/chat/commit/3193ee2d884abc7676feb800faca5c2fa9419924

**Comment by or-else** [comment]
**Problem 3:**

That's by design.

**Comment by or-else** [comment]
**Problem 4-a:**

Confirming, it's a client-side bug.

**Comment by or-else** [comment]
**Problem 4-a:**

Fixed in https://github.com/tinode/tinode-js/commit/c40ebec65cbe52fdf49fe5fb2898f9321cdda560

**Comment by or-else** [comment]
**Problem 4-b**

Should go away after 4-a fix.

**Comment by or-else** [comment]
Everything is done here.

**Comment by sanmiguel2019** [comment]
> **Problem 1:**
> 
> I'm sure that's because you set `DefAcs.Auth` to `N`.
> 

Looks that you are right. I was playing with permissions of the users and probably forgot to set them back to normal.
I know that github is not for questions but I already asked it in the forum but there are mostly questions and only Gene answers. Maybe you or-else can answer this?

Is there a way to create p2p topics with permissions `N` of both users? In my project I want to turn off the possibility to create p2p topics from clients just knowing the user ID. I want to achieve that p2p topics only can create a root user from parent website using GRPC calls.

I tried it in the following way:

1) by default I set permissions for all users to: `auth:"N" anon:"N"`
2) when some user sends a new message from the parent website (creates a p2p topic in Tinode):
- as a root user I set permissions of both users to: `auth:"JRWPS" anon:"N"`
- I create a p2p topic/subscribe between both users
- send message
3) now the topic is created and users can continue chatting from clients. I set permissions of both users (`users` table) back to: `auth:"N" anon:"N"`

Why is does not work? As I see in the database, changing permissions as I described works.  But for some reason both users get the following permissions in the `subscriptions` table:
`N N`
`A A`
This causes `user is banned `error. Why? If in the moment of creating a topic both users have auth:"JRWPS"?

**Comment by sanmiguel2019** [comment]
> Everything here is either done or working as intended. In the future please open separate issues for each of the problems.

Ok, thanks, I will do so.

**Comment by or-else** [comment]
>  In my project I want to turn off the possibility to create p2p topics from clients just knowing the user ID

As I suggested in the answer to this question in the group, the right way to do it is to use `FireHose` plugin.

**Comment by sanmiguel2019** [comment]
Probably but I am not sure that I know how to set it up. The way that described its easier for me but don't know why it doe snot work. Thanks anyway.

**Comment by or-else** [comment]
Looks at the chatbot source.

**Comment by sanmiguel2019** [comment]
Problem with appearing of the newly created topics is fixed now but badges that show the number of new messages still don't appear:

Reproducing:
1. Log in into webapp with 2 different usernames from 2 different devices.
2. As user A search for user B and create a new p2p topic with user B. New p2p topic will appear for user B.
3. As user A send a message to user B. Message will be delivered but new message bagde will not appear for user B. It will appear only when user B clicks to the topic with user B in the contact list.

The same with groups.
1. Create a new group.
2. As user A add user B to the new group.
3. As user A send a message to this group.
4. Message will be delivered but new message badge will not appear for user B.

**Comment by or-else** [comment]
If you found a new problem, please open a new issue.
If you found an old problem re-occurring then please open a new issue and and a reference to the old issue.

This issue is closed. Please do not comment on old closed issues. Open a new one instead.
